node/src/components/consensus/highway_core/state/tests.rs

```diff
@@ -236,6 +297,84 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
+    let a2 = add_unit!(state, ALICE, None; a1, b1, F)?;
+
+    // Now we try to add a2 to a new state.
+    let mut new_state = State::new_test(WEIGHTS, 0);
+
+    // First it should request all of the creator's — Alice's — own units.
+    let a2_swunit = state.wire_unit(&a2, TEST_INSTANCE_ID).unwrap();
+    let res = new_state.compute_panorama(&a2_swunit);
+    assert_eq!(Err(Dependency::UnitBySeqNum(0, ALICE, a2)), res);
+    let a0_swunit = state.wire_unit(&a0, TEST_INSTANCE_ID).unwrap();
+    let a0_panorama = new_state.compute_panorama(&a0_swunit).unwrap();
+    new_state.add_unit(a0_swunit, a0_panorama)?;
+
+    // Once a0 is added, it requests a1. Since a2 knows its predecessor a1 by hash, it doesn't
+    // request it by sequence number.
+    let res = new_state.compute_panorama(&a2_swunit);
+    assert_eq!(Err(Dependency::Unit(a1)), res);
+    let a1_swunit = state.wire_unit(&a1, TEST_INSTANCE_ID).unwrap();
+    let a1_panorama = new_state.compute_panorama(&a1_swunit).unwrap();
+    new_state.add_unit(a1_swunit, a1_panorama)?;
+
+    // With a1 it can start going through its sequence number-based panorama and find b1 missing.
+    // Since it doesn't even have a unit number 0 from Bob, it requests first b0, then b1.
+    let res = new_state.compute_panorama(&a2_swunit);
+    assert_eq!(Err(Dependency::UnitBySeqNum(0, BOB, a2)), res);
+    let b0_swunit = state.wire_unit(&b0, TEST_INSTANCE_ID).unwrap();
+    let b0_panorama = new_state.compute_panorama(&b0_swunit).unwrap();
+    new_state.add_unit(b0_swunit, b0_panorama)?;
+    let res = new_state.compute_panorama(&a2_swunit);
+    assert_eq!(Err(Dependency::UnitBySeqNum(1, BOB, a2)), res);
+    let b1_swunit = state.wire_unit(&b1, TEST_INSTANCE_ID).unwrap();
+
+    // Bob's b1 depends on c0. That should be all we need: Now we can add c0, then b1, then a2.
+    let res = new_state.compute_panorama(&b1_swunit);
+    assert_eq!(Err(Dependency::UnitBySeqNum(0, CAROL, b1)), res);
+
+    // But not so fast! Carol equivocated and we have the wrong unit number 0.
+    let c0_prime_swunit = state.wire_unit(&c0_prime, TEST_INSTANCE_ID).unwrap();
+    let c0_prime_panorama = new_state.compute_panorama(&c0_prime_swunit).unwrap();
+    new_state.add_unit(c0_prime_swunit, c0_prime_panorama)?;
+
+    // Now b1's panorama can be computed but its hash doesn't match!
+    // We need to request the full hash-based panoramas. They contain the right c0 by hash.
+    let res = new_state.compute_panorama(&b1_swunit);
+    assert_eq!(Err(Dependency::UnitWithPanorama(b1)), res);
+    let b1_panorama = state.unit(&b1).panorama.clone();
+    let c0_swunit = state.wire_unit(&c0, TEST_INSTANCE_ID).unwrap();
+    let c0_panorama = new_state.compute_panorama(&c0_swunit).unwrap();
+    new_state.add_unit(c0_swunit, c0_panorama)?;
+
+    // Finally we can add b1 and a2. Since a2 already knows that Carol is faulty, its panorama can
+    // be computed despite the multiple forks.
```

Comment:
Can you add an assertion that verifies that `a2`'s panorama indeed sees Carol as faulty?

Reply:
But that's just the `F` in line 319:
```rust
let a2 = add_unit!(state, ALICE, None; a1, b1, F)?;
```